tools/clang/tools/dxcompiler/dxcpdbutils.cpp

```diff
@@ -390,9 +589,69 @@ struct DxcPdbUtils : public IDxcPdbUtils, public IDxcPixDxilDebugInfoFactory
   virtual HRESULT STDMETHODCALLTYPE GetFullPDB(_COM_Outptr_ IDxcBlob **ppFullPDB) override {
     if (!m_InputBlob)
       return E_FAIL;
+
     if (!ppFullPDB) return E_POINTER;
+
     *ppFullPDB = nullptr;
-    return m_InputBlob.QueryInterface(ppFullPDB);
+
+    // If we are already a full pdb, just return the input blob
+    if (IsFullPDB()) {
+      return m_InputBlob.QueryInterface(ppFullPDB);
+    }
+
+    CComPtr<IDxcCompiler3> pCompiler;
+    IFR(DxcCreateInstance2(m_pMalloc, CLSID_DxcCompiler, __uuidof(IDxcCompiler3), (void **)&pCompiler));
+
+    DxcThreadMalloc TM(m_pMalloc);
+
+    std::vector<const WCHAR *> new_args;
+    for (unsigned i = 0; i < m_Args.size(); i++) {
+      if (m_Args[i] == L"/Qslim_debug" || m_Args[i] == L"-Qslim_debug")
+        continue;
+      new_args.push_back(m_Args[i].c_str());
+    }
+    new_args.push_back(L"-Qfull_debug");
+
+    assert(m_MainFileName.size());
+    if (m_MainFileName.size())
+      new_args.push_back(m_MainFileName.c_str());
+
+    if (m_SourceFiles.empty())
+      return E_FAIL;
+
+    CComPtr<PdbRecompilerIncludeHandler> pIncludeHandler = CreateOnMalloc<PdbRecompilerIncludeHandler>(m_pMalloc);
+    if (!pIncludeHandler)
+      return E_OUTOFMEMORY;
+
+    pIncludeHandler->m_pPdbUtils = this;
+    for (unsigned i = 0; i < m_SourceFiles.size(); i++) {
+      std::wstring NormalizedName = NormalizePath(m_SourceFiles[i].Name.c_str());
+      pIncludeHandler->m_FileMap.insert(std::pair<std::wstring, unsigned>(NormalizedName, i));
+    }
+
+    IDxcBlobEncoding *main_file = m_SourceFiles[0].Content;
+
+    DxcBuffer source_buf = {};
+    source_buf.Ptr = main_file->GetBufferPointer();
+    source_buf.Size = main_file->GetBufferSize();
+    BOOL bEndodingKnown = FALSE;
+    IFR(main_file->GetEncoding(&bEndodingKnown, &source_buf.Encoding));
+
+    CComPtr<IDxcResult> pResult;
+    IFR(pCompiler->Compile(&source_buf, new_args.data(), new_args.size(), pIncludeHandler, __uuidof(IDxcResult), (void **)&pResult));
+
+    CComPtr<IDxcOperationResult> pOperationResult;
+    IFR(pResult.QueryInterface(&pOperationResult));
+
+    HRESULT CompileResult = S_OK;
+    IFR(pOperationResult->GetStatus(&CompileResult));
+    IFR(CompileResult);
+
+    CComPtr<IDxcBlob> pFullPDB;
+    CComPtr<IDxcBlobUtf16> pFullPDBName;
+    IFR(pResult->GetOutput(DXC_OUT_PDB, __uuidof(IDxcBlob), (void **)&pFullPDB, &pFullPDBName));
```

Comment:
Note:
`__uuidof(IDxcBlob), (void **)&pFullPDB`
can just be:
`IID_PPV_ARGS(&pFullPDB)`
Same elsewhere, such as in Compile() earlier.

Reply:
Switched to using the macro.